WebDriverAgentLib/Utilities/FBXPath.m

```diff
@@ -0,0 +1,265 @@
+/**
+ * Copyright (c) 2015-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+#import "FBXPath.h"
+
+#import "FBLogger.h"
+#import "XCAXClient_iOS.h"
+#import "XCTestDriver.h"
+#import "XCTestPrivateSymbols.h"
+#import "XCUIElement.h"
+#import "XCUIElement+FBWebDriverAttributes.h"
+
+const static char *_UTF8Encoding = "UTF-8";
+
+static NSString *const kXMLIndexPathKey = @"private_indexPath";
+static NSString *const topNodeIndexPath = @"top";
+NSString *const XCElementSnapshotInvalidXPathException = @"XCElementSnapshotInvalidXPathException";
+NSString *const XCElementSnapshotXPathQueryEvaluationException = @"XCElementSnapshotXPathQueryEvaluationException";
+
+@implementation FBXPath
+
++ (void)throwException:(NSString *)name forQuery:(NSString *)xpathQuery __attribute__((noreturn))
+{
+  NSString * reason = [NSString stringWithFormat:@"Cannot evaluate results for XPath expression \"%@\"", xpathQuery];
+  @throw [NSException exceptionWithName:name reason:reason userInfo:@{}];
+}
+
++ (NSArray<id<FBElement>> *)findMatchesIn:(id<FBElement>)root xpathQuery:(NSString *)xpathQuery
+{
+  xmlDocPtr doc;
+  
+  xmlTextWriterPtr writer = xmlNewTextWriterDoc(&doc, 0);
+  if (NULL == writer) {
+    [FBLogger logFmt:@"Failed to invoke libxml2>xmlNewTextWriterDoc for XPath query \"%@\"", xpathQuery];
+    [FBXPath throwException:XCElementSnapshotXPathQueryEvaluationException forQuery:xpathQuery];
+    return nil;
+  }
+  NSMutableDictionary *elementStore = [NSMutableDictionary dictionary];
+  int rc = [FBXPath getSnapshotAsXML:root writer:writer elementStore:elementStore];
+  if (rc < 0) {
+    xmlFreeTextWriter(writer);
+    xmlFreeDoc(doc);
+    [FBXPath throwException:XCElementSnapshotXPathQueryEvaluationException forQuery:xpathQuery];
+    return nil;
+  }
+  
+  xmlXPathObjectPtr queryResult = [FBXPath evaluate:xpathQuery document:doc];
+  if (NULL == queryResult) {
+    xmlFreeTextWriter(writer);
+    xmlFreeDoc(doc);
+    [FBXPath throwException:XCElementSnapshotInvalidXPathException forQuery:xpathQuery];
+    return nil;
+  }
+  
+  NSArray *matchingSnapshots = [FBXPath collectMatchingSnapshots:queryResult->nodesetval elementStore:elementStore];
+  xmlXPathFreeObject(queryResult);
+  xmlFreeTextWriter(writer);
+  xmlFreeDoc(doc);
+  if (nil == matchingSnapshots) {
+    [FBXPath throwException:XCElementSnapshotXPathQueryEvaluationException forQuery:xpathQuery];
```

Comment:
Because this error means there are some matches found by direct xpath matching, but XCTest is unable to match these matches to the actual tree, which is exceptional situation.